### PR TITLE
Issue #248: reducing the size of the URI list for bq load

### PR DESCRIPTION
**Pull request by davidrabinowitz**

### CODE REVIEW

File: connector/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryUtil.java [hunk 1]
```diff
@@ -122,4 +128,45 @@ public static String friendlyTableName(TableId tableId) {
...
+   * @param prefixRegex the regex to catch the part before the file index.
+   * @param suffixRegex the regex to catch the part after the file index.
+   */
+  public static List<String> optimizeLoadUriList(
+      List<String> uris, String prefixRegex, String suffixRegex) {
+    Pattern pattern = Pattern.compile("(" + prefixRegex + "\\d*)\\d\\**(" + suffixRegex + ")");
+    ImmutableList.Builder<String> result = ImmutableList.builder();
+    List<String> workingList = uris;
+    Multimap<String, String> trimmedUriMap = HashMultimap.create();
+    while (!workingList.isEmpty()) {
```
Comment (functicons) [inline connector/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryUtil.java:147]: Can we introduce a helper method trimUris which would be cleaner?

while (true) {
  trimmedUris = trimUris(currentUris);
  if (trimmedUris.size() == currentUris.size()) {
     // Couldn't reduce further.
    break;
  }
  currentUri = trimmedUril;
}

File: connector/src/main/java/com/google/cloud/spark/bigquery/SparkBigQueryUtil.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+ */
+package com.google.cloud.spark.bigquery;
+
+import com.google.cloud.bigquery.connector.common.BigQueryUtil;
+
+import java.util.List;
+
+/** Spark related utilities */
+public class SparkBigQueryUtil {
+  public static List<String> optimizeLoadUriListForSpark(List<String> uris) {
```
Comment (functicons) [inline connector/src/main/java/com/google/cloud/spark/bigquery/SparkBigQueryUtil.java:24]: Add a comment about this method.

**Review by functicons — CHANGES_REQUESTED** [review]

File: connector/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryUtil.java (hunk 1 above)
Comment (davidrabinowitz) [reply connector/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryUtil.java:146]: Notice that in case the collection of the mapped URIs contains less than 10 URIs then a wildcard is not being used and they are added to the result list. Moving the entire logic into the helper method would mean to process those several times in case of a large result. I took your advice for cleaning the method and moved the trimming part to another method.

File: connector/src/test/java/com/google/cloud/spark/bigquery/OptimizeLoadUriListTest.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+  private List<String> input;
+  private List<String> expected;
+
+  public OptimizeLoadUriListTest(List<String> input, List<String> expected) {
+    this.input = input;
+    this.expected = expected;
+  }
+
+  @Parameterized.Parameters(name = "{index}: Should get {1}")
+  public static Iterable<Object[]> data() {
```
Comment (functicons) [inline connector/src/test/java/com/google/cloud/spark/bigquery/OptimizeLoadUriListTest.java:41]: Can we add edge cases for input list of size 0 and 1?

**Review by functicons — APPROVED** [review]

File: connector/src/test/java/com/google/cloud/spark/bigquery/OptimizeLoadUriListTest.java (hunk 3 above)
Comment (davidrabinowitz) [reply connector/src/test/java/com/google/cloud/spark/bigquery/OptimizeLoadUriListTest.java:41]: Added, thanks
